subscriber/src/daily_operations.rs

```diff
@@ -0,0 +1,227 @@
+        if progress.current_index == 0 {
+            progress.service_index = service_index;
+            progress.current_index = *user_index;
+            progress.additional_data_index = 0;
+        } else {
+            require!(
+                progress.service_index == service_index,
+                "Another operation is in progress"
+            );
+        }
+
+        let fees_contract_address = self.fees_contract_address().get();
+        let energy_threshold = self.energy_threshold().get();
+        let service_info = self.service_info().get().get(service_index);
+        let current_epoch = self.blockchain().get_block_epoch();
+        let additional_data_len = additional_data.len();
+
+        let run_result = self.run_while_it_has_gas(GAS_TO_SAVE_PROGRESS, || {
+            if progress.additional_data_index >= additional_data_len
+                || progress.current_index > total_users
+            {
+                return STOP_OP;
+            }
+
+            let user_data = additional_data.get(*user_index);
+            progress.additional_data_index += 1;
+
+            let user_id = users_mapper.get_by_index(progress.current_index);
+            let opt_user_address = self
+                .user_id()
+                .get_address_at_address(&fees_contract_address, user_id);
+            let user_address = match opt_user_address {
+                Some(address) => address,
+                None => {
+                    users_mapper.swap_remove(&user_id);
+                    total_users -= 1;
+                    return CONTINUE_OP;
+                }
+            };
+
+            progress.current_index += 1;
+
+            let subscription_type = self.subscription_type(user_id, service_index).get();
+            let next_action_epoch =
+                self.get_next_action_epoch(user_id, service_index, subscription_type);
+            if next_action_epoch > current_epoch {
+                return CONTINUE_OP;
+            }
+
+            let subtract_result = self.subtract_payment(
+                &user_address,
+                user_id,
+                &energy_threshold,
+                service_info.payment_type.clone(),
+                fees_contract_address.clone(),
+            );
+            if subtract_result.is_err() {
+                return CONTINUE_OP;
+            }
+
+            let action_results = SC::perform_action(
+                self,
+                user_address.clone(),
+                user_id,
+                &service_info,
+                user_data.borrow(),
+            );
+            if action_results.is_err() {
+                return CONTINUE_OP;
+            }
+
+            // return funds if it didn't work? - discuss
+
+            let interpreted_results = unsafe { action_results.unwrap_unchecked() };
+            if let Some(new_token) = interpreted_results.opt_new_token {
+                self.save_new_token(user_id, new_token);
```

Comment:
we do not have new token on claimBoostedRewards. and we can make claimBoostedRewards without the need of sending any tokens.